### PR TITLE
Change the way a candidate's 'current application' is determined

### PR DESCRIPTION
**Pull request by TobyRet**
## Context

We've noticed some tricky bugs occur in code that relies on the `current_application` method in the Candidate model (see https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/app/models/candidate.rb#L26)

It's possible that `application_forms.last` may not always return what is thought to be the latest application form if prior to this, the application_forms  array has been mutated in some way.

Instead the safer option would be to order by created date 
 `application_forms.order(:created_at).last` 

## Replicating the bug

In the console find a candidate with multiple application forms (or create them)

```
# previously the '#current_application' returns the application form with the largest id (or creates and application form and returns it)

def current_application
  application_form = application_forms.last
  ...
end


# At this point the current_application id is as expected

candidate.current_application.id = x

# Somewhere else in the code, we try to determine a candidate's previous applications (as an array). Note how the 'current_application' is subtracted from all applications

previous_applications = candidate.application_forms - [ candidate.current_application ]

# But now the current_application id is  x - 1
candidate.current_application.id = x

```

## Changes proposed in this pull request

- Determine the current application by sorting all application forms by created date and return the most recent application form
- The above change has impacted pre-filling of blank applications (in Sandbox). Blank applications are destroyed in order to make [the system test](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/spec/system/candidate_interface/signup_and_signin/candidate_signs_in_and_prefills_application_in_sandbox_spec.rb) green. See https://github.com/DFE-Digital/apply-for-teacher-training/pull/4418/commits/41d60d0ffd128c8d43449e25aa829b3740ae9a50 for context

## Link to Trello card

https://trello.com/c/tOE6GT5l/3247-change-the-way-a-candidates-current-application-is-determined

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] API release notes have been updated if necessary
- [ ] This code does not rely on the addition/removal of Azure config environment variables in the same Pull Request
- [ ] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

**Review by davidgisbey — APPROVED** [review]

**Comment by TobyRet** [comment]
@stevehook - I'm getting a single failure on the `mailers_controller_spec` 

mailer_error

A 'candidate` is [passed into the method ](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/app/mailers/candidate_mailer.rb#L369) but upon inspecting the Candidate `COUNT` is 0.

**Comment by stevehook** [comment]
> @stevehook - I'm getting a single failure on the `mailers_controller_spec`. Do you know what might be going on here?
> 
> mailer_error
> 
> A 'candidate' is [passed into the method](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/app/mailers/candidate_mailer.rb#L369) but upon inspecting the Candidate `COUNT` is 0.

So I think this is happening just because `candidate` in this test is created with `build_stubbed`. It's a fake - it has an `id` but isn't persisted. So when `current_application` is called here - https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/app/mailers/candidate_mailer.rb#L369, `Candidate#current_application` calls `application_forms.order(:created_at).last` and because of the `order` (and some internals to do with FactoryBot stubbing that I don't understand) it has to run a query (rather than just returning the `application_forms` in memory) and ends up trying to persist the application forms first.

**Comment by TobyRet** [comment]
> So I think this is happening just because `candidate` in this test is created with `build_stubbed`. It's a fake - it has an `id` but isn't persisted. So when `current_application` is called here - https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/app/mailers/candidate_mailer.rb#L369, `Candidate#current_application` calls `application_forms.order(:created_at).last` and because of the `order` (and some internals to do with FactoryBot stubbing that I don't understand) it has to run a query (rather than just returning the `application_forms` in memory) and ends up trying to persist the application forms first.
> 
> `application_forms.sort_by(&:created_at).last` works because it's not hitting the DB but that doesn't seem like the right thing to do.

Yep looks like it is an issue with `build_stubbed` - where the specs are broken we need to persist the candidate and associations. Have updated specs and now looks ok. Thanks @stevehook

File: app/controllers/candidate_interface/prefill_application_form_controller.rb [hunk 1]
```diff
@@ -37,5 +40,17 @@ def prefill_candidate_application_form
...
+      blank_application.destroy if blank_application?
+    end
+
+    def blank_application?
+      blank_application.created_at == blank_application.updated_at
+    end
+
+    def blank_application
+      @blank_application ||= current_candidate.application_forms.first
+    end
```
Comment (TobyRet) [inline app/controllers/candidate_interface/prefill_application_form_controller.rb:54]: @benswannack - I've added a guard clause to ensure we only destroy applications that are actually blank. What do you think?

Not sure if this logic should sit here or somewhere else though ... Could potentially create a service but is that overkill? 🤔

Comment (TobyRet) [reply app/controllers/candidate_interface/prefill_application_form_controller.rb:54]: Maybe `#blank_application?` could move to the model 🤔

Comment (duncanjbrown) [reply app/controllers/candidate_interface/prefill_application_form_controller.rb:54]: There's already a `#blank_application?` method on `ApplicationForm`

**Comment by duncanjbrown** [comment]
This should solve https://trello.com/c/fT2Fvq3O/3531-on-sandbox-we-ask-the-would-you-like-to-prefill-this-application-question-when-the-candidate-has-an-application-in-progress as well?

File: app/controllers/candidate_interface/prefill_application_form_controller.rb (hunk 1 above)
Comment (TobyRet) [reply app/controllers/candidate_interface/prefill_application_form_controller.rb:54]: Ah yes - there is! I'll fix

**Review by benswannack — APPROVED** [review]
This looks good to me! (and will fix the ticket Duncan has linked as a bonus :) )

**Comment by TobyRet** [comment]
> This should solve https://trello.com/c/fT2Fvq3O/3531-on-sandbox-we-ask-the-would-you-like-to-prefill-this-application-question-when-the-candidate-has-an-application-in-progress as well?

@duncanjbrown - I was not able to replicate the bug you listed in the card on master.
I did the following ...
1.  signed up for an account
2.  selected 'pre-fill application'
3. signed out
4. was taken straight to my pre-filled application

File: .env.development [hunk 2]
```diff
@@ -4,3 +4,4 @@ HOSTING_ENVIRONMENT_NAME=development
 STATE_CHANGE_SLACK_URL=
 GOOGLE_ANALYTICS_APPLY=UA-APPLY-XX
 GOOGLE_ANALYTICS_MANAGE=UA-MANAGE-XX
+SANDBOX=true
```
Comment (benswannack) [inline .env.development:7]: Is this intentional?

Comment (TobyRet) [reply .env.development]: Woops nope - have reverted

**Pull request merged**